llmfoundry/tokenizers/tiktoken.py

```diff
@@ -0,0 +1,276 @@
+            eos_token (Optional[str], optional): The eos token. Defaults to '<|endoftext|>'.
+            bos_token (Optional[str], optional): The bos token. Defaults to '<|endoftext|>'.
+            pad_token (Optional[str], optional): The pad token. Defaults to None.
+        """
+        try:
+            import tiktoken
+        except:
+            raise ImportError(
+                'You need to install tiktoken to use TiktokenTokenizerWrapper.')
+
+        if model_name is not None and encoding_name is not None:
+            raise ValueError(
+                'You need to specify either model_name or encoding_name, not both.'
+            )
+
+        self.model_name = model_name
+        self.encoding_name = encoding_name
+
+        if self.model_name is not None:
+            self.encoding = tiktoken.encoding_for_model(  # type: ignore (thirdParty)
+                self.model_name)
+        elif self.encoding_name is not None:
+            self.encoding = tiktoken.get_encoding(  # type: ignore (thirdParty)
+                self.encoding_name)
+        else:
+            raise ValueError(
+                'You need to specify either model_name or encoding_name.')
+
+        self.add_bos_token = add_bos_token
+
+        super().__init__(model_name=model_name,
+                         encoding_name=encoding_name,
+                         add_bos_token=add_bos_token,
+                         unk_token=unk_token,
+                         eos_token=eos_token,
+                         bos_token=bos_token,
+                         pad_token=pad_token,
+                         **kwargs)
+
+    @property
+    def vocab_size(self):
+        """Returns vocab size."""
+        return self.encoding.n_vocab
+
+    @property
+    def is_fast(self) -> bool:
+        return False
+
+    def get_vocab(self):
+        """Returns vocab as a dict."""
+        vocab = {}
+        for i in range(self.vocab_size):
+            try:
+                # need to try this first, so that we get a proper KeyError,
+                # otherwise it crashes in the rust code
+                _ = self.encoding.decode_single_token_bytes(i)
+                vocab[self.encoding.decode([i])] = i
+            except KeyError:
+                pass
+
+        return vocab
+
+    def _tokenize(self, text: str):
+        """Returns a tokenized string.
+
+        Note: We have slightly redefined the expected contract between this method and
+        the _convert_token_to_id method. Normally, this method turns a string, into a list of strings,
+        and then the _convert_token_to_id method turns that list of strings into a list of integers.
+        However, not all vocab indices can be decoded into a string, so instead we just return the integers
+        from this function, and have adjusted the _convert_token_to_id method to handle integers as well as strings.
+        The only use of _tokenize that I could find was in this way, so this _should_ be safe.
```

Comment:
Is this because special chars cause problems? A little weird to change

Reply:
its because of
```
In [1]: import tiktoken

In [2]: gpt4t = tiktoken.encoding_for_model('gpt-4')

In [3]: gpt4t.encode('Å')
Out[3]: [127, 227]

In [4]: gpt4t.decode([127, 227])
Out[4]: 'Å'

In [5]: gpt4t.decode([127])
Out[5]: '�'

In [6]: gpt4t.decode([227])
Out[6]: '�'
```